### ISSUE TITLE
btcwallet misses sometimes transactions present in btcd

### ISSUE THREAD

**Issue by arnuschky** [btcsuite/btcwallet#259]
I am comparing btcwallet and bitcoin core in-system at the moment.

I am (sometimes) seeing unspents listed by btcwallet that have been spent already. The transaction that as spent it is always known to btcd, but not to btcwallet.

First time, I waited a few minutes, but the transaction wasn't propagated to btcwallet. Only fix I am aware of is to rescan the wallet, which picks up the missing transaction.

What could I do to debug further? Any hints?

**Comment by arnuschky** [comment]
To clarify: `getrawtransaction` returns the tx, while `gettransaction` doesn't and `listunspents` still lists the unspent output.

**Comment by jrick** [comment]
Odd.  What kind of scenario are you testing wallet in?  Do you see any pattern to the missing transactions?

**Comment by jrick** [comment]
Oh, I see a bug.  It removes transactions from one extra block when a reorg is notified.  Can you find the reorganize in the btcd logs, and check to see if the missing transactions were from the block just before the reorganize?

**Comment by jrick** [comment]
If that is the bug, I'm not certain why the previous outputs would still be marked unspent.  The spending transactions should have been moved around in wtxmgr to be handled like other unmined transactions, but the previous outputs should still be considered spent.

**Comment by arnuschky** [comment]
I can't find anything about this block in the logs other that they've been synced (they are old, below 100000). No chain reorg observable.

Even weirder, I now have the same issue with a transaction down the line. Before, I had unspent A because tx B was unknown. Now I have unspent B because tx C hasn't been recognized as spending. This time, however, `gettransaction` knows C, which seems weird. 

I can recreate an orderly state by rescanning, but I figured I don't disturb the evidence for now. I am not sure which conditions trigger this problem; it might be that my test setup is responsible for this, even though I don't see how.

BTW, I have been observing `[ERR] WLLT: Could not lock wallet: address manager is locked`, but couldn't make a definitive link with the above bug.

**Comment by arnuschky** [comment]
Just realized that I am not running the latest version, mine is ~10 days old. I'll leave it for now, unless there's a good reason to update.

**Comment by jrick** [comment]
Wallet got some pretty significant changes over the past week.  We're currently getting some remaining things in order so we can roll out a release soon.  I also discovered quite a few reorg bugs while implementing wtxmgr (the replacement for the older txstore package you are using) which were probably carried over from txstore logic.

**Comment by arnuschky** [comment]
I see. I upgraded go to 1.3.3 and btc-suite to latest. Currently rescanning, I will start testing again once this is done.

A first feedback: rescan is about 10x slower on the new version. But I am getting debug output, yay!

**Comment by davecgh** [comment]
As a point of reference, while btcd still officially supports Go 1.3.3, it is pretty old at this point.  If you're upgrading anyways, I'd recommend Go 1.4.2.

The Go 1.4.x run-time added precise GC among other things that provide lower heap usage and faster execution.

**Comment by arnuschky** [comment]
Ok, thanks for the hint. Not very knowledgable in terms of go, all my distributions deliver <=1.1 so I felt pretty spiffy with 1.3.3.

Rescan seems to be stuck, or at least slowed down to a crawl (running since 3 hours). This is clearly a regression, this wallet took < 5 minutes to rescan before. Is anyone else experiencing this?

Is there a new data structure that becomes exponentially slower when inserting transactions? Ironically, this is exactly what we've observed in bitcoin core (and why we need to switch :) )

**Comment by jrick** [comment]
Can you tell whether it's wallet slowing down, or btcd?

**Comment by arnuschky** [comment]
It's the wallet. btcd is fine, but there's always one thread of btcwallet that's blocked (process state D, uninterruptable sleep = IO). iotop shows ~700-900kB/s IO.

**Comment by jrick** [comment]
Thanks, sounds like bolt writes are very slow.  We're talking right now about throwing a WAL on top of the DB to increase write throughput.

**Comment by arnuschky** [comment]
I could not replicate the missing transaction bug with the new version based on wtxmgr.

I am therefore closing this in favor of#269.
